app/vmctl/opentsdb.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/app/vmctl/opentsdb"
+	"github.com/VictoriaMetrics/VictoriaMetrics/app/vmctl/vm"
+	"github.com/cheggaaa/pb/v3"
+)
+
+type otsdbProcessor struct {
+	oc      *opentsdb.Client
+	im      *vm.Importer
+	otsdbcc int
+	vmcc    int
+}
+
+type queryObj struct {
+	Series    opentsdb.Meta
+	Rt        opentsdb.Retention
+	Tr        opentsdb.TimeRange
+	StartTime int64
+}
+
+func newOtsdbProcessor(oc *opentsdb.Client, im *vm.Importer, otsdbcc int, vmcc int) *otsdbProcessor {
+	if otsdbcc < 1 {
+		otsdbcc = 1
+	}
+	if vmcc < 1 {
+		vmcc = 1
+	}
+	return &otsdbProcessor{
+		oc:      oc,
+		im:      im,
+		otsdbcc: otsdbcc,
+		vmcc:    vmcc,
+	}
+}
+
+func (op *otsdbProcessor) run(silent bool) error {
+	log.Println("Loading all metrics from OpenTSDB for filters: ", op.oc.Filters)
+	var metrics []string
+	for _, filter := range op.oc.Filters {
+		q := fmt.Sprintf("%s/api/suggest?type=metrics&q=%s&max=%d", op.oc.Addr, filter, op.oc.Limit)
+		m, err := op.oc.FindMetrics(q)
+		if err != nil {
+			return fmt.Errorf("metric discovery failed for %q: %s", q, err)
+		}
+		metrics = append(metrics, m...)
+	}
+	if len(metrics) < 1 {
+		return fmt.Errorf("found no timeseries to import with filters %q", op.oc.Filters)
+	}
+
+	question := fmt.Sprintf("Found %d metrics to import. Continue?", len(metrics))
+	if !silent && !prompt(question) {
+		return nil
+	}
+
+	seriesCh := make(chan queryObj)
+	errCh := make(chan error)
+	startTime := time.Now().Unix()
+	queryRanges := 0
+	for _, rt := range op.oc.Retentions {
+		queryRanges += len(rt.QueryRanges)
+	}
+	for _, metric := range metrics {
+		log.Println(fmt.Sprintf("Starting work on %s", metric))
+		serieslist, err := op.oc.FindSeries(metric)
+		if err != nil {
+			return fmt.Errorf("couldn't retrieve series list for %s : %s", metric, err)
+		}
+		bar := pb.StartNew(len(serieslist) * queryRanges)
+		var wg sync.WaitGroup
+		wg.Add(op.otsdbcc)
+		for i := 0; i < op.otsdbcc; i++ {
+			go func() {
+				defer wg.Done()
+				for s := range seriesCh {
+					if err := op.do(s); err != nil {
+						errCh <- fmt.Errorf("couldn't retrieve series for %s : %s", metric, err)
+						return
+					}
+					bar.Increment()
+				}
+			}()
+		}
+		// log.Println(fmt.Sprintf("Found %d series for %s", len(serieslist), metric))
+		for _, series := range serieslist {
+			select {
+			case otsdbErr := <-errCh:
+				return fmt.Errorf("opentsdb error: %s", otsdbErr)
+			default:
+				for _, rt := range op.oc.Retentions {
+					for _, tr := range rt.QueryRanges {
+						seriesCh <- queryObj{
```

Comment:
Could be a deadlock here. Consider the case when `op.do(s)` returns an error and you still have `rt.QueryRanges` not drained. In such case both goroutines will stuck in attempt to send into channels:
1.  No one reads from `errCh` since main goroutine is trying to push into `seriesCh`
2. No one reads from `seriesCh` since concurrent goroutine is trying to push into `errCh`

Reply:
I'm now reading from errCh (and vm insert error channel) consistently with adding new series...does that solve this problem?